d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -534,34 +581,67 @@ private void handleResourceRemoval(Collection<String> removedResources, Resource
     {
       return;
     }
+
+    Map<String, ResourceSubscriber> subscribers = getResourceSubscriberMap(type);
+    WildcardResourceSubscriber wildcardSubscriber = getWildcardResourceSubscriber(type);
     for (String resourceName : removedResources)
     {
       _xdsClientJmx.incrementResourceNotFoundCount();
       _log.warn("Received response that {} {} was removed", type, resourceName);
-      ResourceSubscriber subscriber = getResourceSubscriberMap(type).get(resourceName);
+
+      ResourceSubscriber subscriber = subscribers.get(resourceName);
       if (subscriber != null)
       {
         subscriber.onRemoval();
       }
+
+      if (wildcardSubscriber != null)
+      {
+        wildcardSubscriber.onRemoval(resourceName);
+      }
     }
-    WildcardResourceSubscriber wildcardSubscriber = getWildcardResourceSubscriber(type);
-    if (wildcardSubscriber != null)
+  }
+
+  // If the nonce indicates that this is the last chunk of the response, notify the wildcard subscriber.
+  private void notifyOnLastChunk(DiscoveryResponseData response)
+  {
+    ResourceType type = response.getResourceType();
+    WildcardResourceSubscriber wildcardResourceSubscriber = getWildcardResourceSubscriber(type);
+    if (wildcardResourceSubscriber == null)
     {
-      removedResources.forEach(wildcardSubscriber::onRemoval);
+      return;
     }
-  }
 
+    int remainingChunks;
+    try
+    {
+      byte[] bytes = Hex.decodeHex(response.getNonce().toCharArray());
+      ByteBuffer bb = ByteBuffer.wrap(Arrays.copyOfRange(bytes, 8, 12));
+      remainingChunks = bb.getInt();
+    }
+    catch (Exception e)
+    {
+      _log.warn("Failed to decode nonce: {}", response.getNonce(), e);
+      remainingChunks = -1;
+    }
+
+    if (remainingChunks == 0)
+    {
+      _log.info("Notifying wildcard subscriber of type {} for the end of response chunks.", type);
+      wildcardResourceSubscriber.onAllResourcesProcessed();
+    }
```

Comment:
Something about this feels wrong. For example, if you fail to deserialize the nonce for whatever reason, `onAllResourcesProcessed` may never be called and things will hang forever (for example if you're talking to a really old version of the observer or something weird like that). I think you should actually reverse this, where you *always* call `onAllResourcesProcessed` _unless_ you get explicit signal that `remainingChunks > 0`. Functionally it's the same, but it'll work better if the Observer doesn't send the chunk suffix. This way it's just "best effort" where we try to handle chunks if we can, but fire `onAllResourcesProcessed` early otherwise, which is what you would have to do if you weren't using `diderot` anyway

Reply:
good call! fixed.